src/OpenTelemetry.Api/Context/Propagation/TraceContextPropagator.cs

```diff
@@ -230,18 +237,72 @@ internal static bool TryExtractTracestate(string[] tracestateCollection, out str
 
             if (tracestateCollection != null)
             {
-                var result = new StringBuilder();
+                List<string> listMembers = new List<string>();
```

Comment:
I think in theory we could do away with this `List<string>` allocation. How that would work: Convert the `tracestateCollection[i]` you are working on to a `Span<char>`. Then you can `Slice` up that `Span` on the position of `,` chars. Work on the slices one-by-one, with an `int` counter. If you get to 33, return false. Add to the `StringBuilder` + `HashSet<string>` as you go. In the case of too many keys we would waste processing the first 32, but that should be an edge case? Assuming most tracestates will be valid.

Reply:
I don't get the benefit of this approach. Is it to fail fast when there are too many list members?

If that's the case I could lift the 32 count limit check inside of the for loop of adding to `listMembers` list. Unless there are too many list members in the first string of `tracestateCollection` (note that this is an array of strings where each element is a comma-delimited string: `["foo=1,bar=2", "rojo=1,congo=2", "baz=3"]`, as shown in [test case](https://github.com/w3c/trace-context/blob/master/test/test.py#L710)). In this case this approach would be the same as mine, just replacing `string.Split` with `Span<char>.Slice`.

Reply:
I tested with case `"foo=1,bar=2", "rojo=1,congo=2", "baz=3"`. Using Span and Slice is indeed faster:

|                       Method | tracestateCollection |     Mean |     Error |    StdDev |   Median | Ratio | RatioSD |
|----------------------------- |--------------------- |---------:|----------:|----------:|---------:|------:|--------:|
|         TryExtractTracestate |            String[3] | 1.861 us | 0.0468 us | 0.1328 us | 1.824 us |  1.00 |    0.00 |
| TryExtractTracestateWithSpan |            String[3] | 1.203 us | 0.0238 us | 0.0503 us | 1.199 us |  0.65 |    0.04 |

Reply:
@xiang17 Awesome! Thanks for doing that. Add `[MemoryDiagnoser]` to the benchmark and then do a test with 32 KVPs and you should see even more benefit. My theory was this Span/Slice logic would allow us to get rid of the `List<string>` that saves at least one allocation. But as the list grows, we save more allocations & copying from one buffer to another. Whatever the initial capacity is (8?) it will double each time it reaches that and then copy items to the new one. Could probably have just used 32 as the initial capacity to `List<string>` but if it is faster and saves on allocations, great 😄 Split also has to do some allocation + copying and this removes those too.

Reply:
PS: There's a [runtime issue](https://github.com/dotnet/runtime/issues/934) that would make this type of thing a bit easier.

Reply:
For later reference. I tested with `[MemoryDiagnoser]` tag.


Input: ```{
                "bar01=01,bar02=02,bar03=03,bar04=04,bar05=05,bar06=06,bar07=07,bar08=08,bar09=09,bar10=10",
                "bar11=11,bar12=12,bar13=13,bar14=14,bar15=15,bar16=16,bar17=17,bar18=18,bar19=19,bar20=20",
                "bar21=21,bar22=22,bar23=23,bar24=24,bar25=25,bar26=26,bar27=27,bar28=28,bar29=29,bar30=30",
                "bar31=31,bar32=32"}```


|                       Method | tracestateCollection |     Mean |     Error |    StdDev | Ratio |  Gen 0 | Gen 1 | Gen 2 | Alloca
|----------------------------- |--------------------- |---------:|----------:|----------:|------:|-------:|------:|------:|-------
|         TryExtractTracestate |            String[4] | 8.372 us | 0.1168 us | 0.1036 us |  1.00 | 1.8921 |     - |     - |   7.77
| TryExtractTracestateWithSpan |            String[4] | 6.529 us | 0.0887 us | 0.0787 us |  0.78 | 1.4114 |     - |     - |   5.78


Code


```csharp
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using System.Text;
using BenchmarkDotNet;
using BenchmarkDotNet.Attributes;

namespace gh_1668
{
    [MemoryDiagnoser]
    public class Benchmarks
    {
        private static readonly char[] OptionalWhiteSpaceCharacters = new char[] { ' ', '\t' };

        // The following length limits are from Trace Context v1 https://www.w3.org/TR/trace-context-1/#key
        private static readonly int TraceStateKeyMaxLength = 256;
        private static readonly int TraceStateKeyTenantMaxLength = 241;
        private static readonly int TraceStateKeyVendorMaxLength = 14;
        private static readonly int TraceStateValueMaxLength = 256;

        // public field
        [ParamsSource(nameof(InputValues))]
        public string[] tracestateCollection;

        public static IEnumerable<string[]> InputValues() => new[]
        {
            new string[] {
                "bar01=01,bar02=02,bar03=03,bar04=04,bar05=05,bar06=06,bar07=07,bar08=08,bar09=09,bar10=10",
                "bar11=11,bar12=12,bar13=13,bar14=14,bar15=15,bar16=16,bar17=17,bar18=18,bar19=19,bar20=20",
                "bar21=21,bar22=22,bar23=23,bar24=24,bar25=25,bar26=26,bar27=27,bar28=28,bar29=29,bar30=30",
                "bar31=31,bar32=32"}
        };

        [Benchmark(Baseline = true)]
        public bool TryExtractTracestate()
        {
            string tracestateResult = string.Empty;

            if (tracestateCollection != null)
            {
                List<string> listMembers = new List<string>();
                for (int i = 0; i < tracestateCollection.Length; ++i)
                {
                    // test_tracestate_multiple_headers_different_keys
                    listMembers.AddRange(tracestateCollection[i].Split(','));
                }

                if (listMembers.Count > 32)
                {
                    // https://github.com/w3c/trace-context/blob/master/spec/20-http_request_header_format.md#list
                    // test_tracestate_member_count_limit
                    return false;
                }

                var keySet = new HashSet<string>();
                var result = new StringBuilder();
                for (int i = 0; i < listMembers.Count; ++i)
                {
                    // https://github.com/w3c/trace-context/blob/master/spec/20-http_request_header_format.md#tracestate-header-field-values
                    if (string.IsNullOrEmpty(listMembers[i]))
                    {
                        // Empty and whitespace - only list members are allowed.
                        // Vendors MUST accept empty tracestate headers but SHOULD avoid sending them.
                        continue;
                    }

                    // Spaces and horizontal tabs surrounding list-members are ignored.
                    var listMember = listMembers[i].Trim(OptionalWhiteSpaceCharacters);
                    int keyLength = listMember.IndexOf('=');
                    if (keyLength == listMember.Length || keyLength == -1)
                    {
                        // Missing key or value in tracestate
                        return false;
                    }

                    var key = listMember.Substring(0, keyLength);
                    if (!ValidateKey(key))
                    {
                        // test_tracestate_key_illegal_characters in https://github.com/w3c/trace-context/blob/master/test/test.py
                        // test_tracestate_key_length_limit
                        // test_tracestate_key_illegal_vendor_format
                        return false;
                    }

                    var value = listMember.Substring(keyLength + 1);
                    if (!ValidateValue(value))
                    {
                        // test_tracestate_value_illegal_characters
                        return false;
                    }

                    // ValidateKey() call above has ensured the key does not contain upper case letters.
                    if (!keySet.Add(key))
                    {
                        // test_tracestate_duplicated_keys
                        return false;
                    }

                    if (result.Length > 0)
                    {
                        result.Append(',');
                    }

                    result.Append(listMember);
                }

                tracestateResult = result.ToString();
            }

            return true;
        }

        [Benchmark]
        public bool TryExtractTracestateWithSpan()
        {
            string tracestateResult = string.Empty;

            if (tracestateCollection != null)
            {
                var keySet = new HashSet<string>();
                var result = new StringBuilder();
                for (int i = 0; i < tracestateCollection.Length; ++i)
                {
                    var tracestate = tracestateCollection[i].AsSpan();
                    int begin = 0;
                    while (begin < tracestate.Length) {
                        int length = tracestate.Slice(begin).IndexOf(',');
                        ReadOnlySpan<char> listMember;
                        if (length != -1)
                        {
                            listMember = tracestate.Slice(begin, length).Trim();
                            begin += length + 1;
                        }
                        else
                        {
                            listMember = tracestate.Slice(begin).Trim();
                            begin = tracestate.Length;
                        }

                        // https://github.com/w3c/trace-context/blob/master/spec/20-http_request_header_format.md#tracestate-header-field-values
                        if (listMember.IsEmpty)
                        {
                            // Empty and whitespace - only list members are allowed.
                            // Vendors MUST accept empty tracestate headers but SHOULD avoid sending them.
                            continue;
                        }

                        if (keySet.Count >= 32)
                        {
                            // https://github.com/w3c/trace-context/blob/master/spec/20-http_request_header_format.md#list
                            // test_tracestate_member_count_limit
                            return false;
                        }


                        int keyLength = listMember.IndexOf('=');
                        if (keyLength == listMember.Length || keyLength == -1)
                        {
                            // Missing key or value in tracestate
                            return false;
                        }

                        var key = listMember.Slice(0, keyLength);
                        if (!ValidateKey(key))
                        {
                            // test_tracestate_key_illegal_characters in https://github.com/w3c/trace-context/blob/master/test/test.py
                            // test_tracestate_key_length_limit
                            // test_tracestate_key_illegal_vendor_format
                            return false;
                        }

                        var value = listMember.Slice(keyLength + 1);
                        if (!ValidateValue(value))
                        {
                            // test_tracestate_value_illegal_characters
                            return false;
                        }

                        // ValidateKey() call above has ensured the key does not contain upper case letters.
                        if (!keySet.Add(key.ToString()))
                        {
                            // test_tracestate_duplicated_keys
                            return false;
                        }

                        if (result.Length > 0)
                        {
                            result.Append(',');
                        }

                        result.Append(listMember.ToString());
                    }

                }

                tracestateResult = result.ToString();
            }

            return true;
        }


        private static byte HexCharToByte(char c)
        {
            if ((c >= '0') && (c <= '9'))
            {
                return (byte)(c - '0');
            }

            if ((c >= 'a') && (c <= 'f'))
            {
                return (byte)(c - 'a' + 10);
            }

            throw new ArgumentOutOfRangeException(nameof(c), c, $"Invalid character: {c}.");
        }

        private static bool ValidateKey(string key)
        {
            // This implementation follows Trace Context v1 which has W3C Recommendation.
            // https://www.w3.org/TR/trace-context-1/#key
            // It will be slightly differently from the next version of specification in GitHub repository.

            // There are two format for the key. The length rule applies to both.
            if (key.Length <= 0 || key.Length > TraceStateKeyMaxLength)
            {
                return false;
            }

            // The first format:
            // key = lcalpha 0*255( lcalpha / DIGIT / "_" / "-"/ "*" / "/" )
            // lcalpha = % x61 - 7A; a - z
            // (There is an inconsistency in the expression above and the description in note.
            // Here is following the description in note:
            // "Identifiers MUST begin with a lowercase letter or a digit.")
            if (!IsLowerAlphaDigit(key[0]))
            {
                return false;
            }

            int tenantLength = -1;
            for (int i = 1; i < key.Length; ++i)
            {
                char ch = key[i];
                if (ch == '@')
                {
                    tenantLength = i;
                    break;
                }

                if (!(IsLowerAlphaDigit(ch)
                    || ch == '_'
                    || ch == '-'
                    || ch == '*'
                    || ch == '/'))
                {
                    return false;
                }
            }

            if (tenantLength == -1)
            {
                // There is no "@" sign. The key follow the first format.
                return true;
            }

            // The second format:
            // key = (lcalpha / DIGIT) 0 * 240(lcalpha / DIGIT / "_" / "-" / "*" / "/") "@" lcalpha 0 * 13(lcalpha / DIGIT / "_" / "-" / "*" / "/")
            if (tenantLength == 0 || tenantLength > TraceStateKeyTenantMaxLength)
            {
                return false;
            }

            int vendorLength = key.Length - tenantLength - 1;
            if (vendorLength == 0 || vendorLength > TraceStateKeyVendorMaxLength)
            {
                return false;
            }

            for (int i = tenantLength + 1; i < key.Length; ++i)
            {
                char ch = key[i];
                if (!(IsLowerAlphaDigit(ch)
                    || ch == '_'
                    || ch == '-'
                    || ch == '*'
                    || ch == '/'))
                {
                    return false;
                }
            }

            return true;
        }

        private static bool ValidateKey(ReadOnlySpan<char> key)
        {
            // This implementation follows Trace Context v1 which has W3C Recommendation.
            // https://www.w3.org/TR/trace-context-1/#key
            // It will be slightly differently from the next version of specification in GitHub repository.

            // There are two format for the key. The length rule applies to both.
            if (key.Length <= 0 || key.Length > TraceStateKeyMaxLength)
            {
                return false;
            }

            // The first format:
            // key = lcalpha 0*255( lcalpha / DIGIT / "_" / "-"/ "*" / "/" )
            // lcalpha = % x61 - 7A; a - z
            // (There is an inconsistency in the expression above and the description in note.
            // Here is following the description in note:
            // "Identifiers MUST begin with a lowercase letter or a digit.")
            if (!IsLowerAlphaDigit(key[0]))
            {
                return false;
            }

            int tenantLength = -1;
            for (int i = 1; i < key.Length; ++i)
            {
                char ch = key[i];
                if (ch == '@')
                {
                    tenantLength = i;
                    break;
                }

                if (!(IsLowerAlphaDigit(ch)
                    || ch == '_'
                    || ch == '-'
                    || ch == '*'
                    || ch == '/'))
                {
                    return false;
                }
            }

            if (tenantLength == -1)
            {
                // There is no "@" sign. The key follow the first format.
                return true;
            }

            // The second format:
            // key = (lcalpha / DIGIT) 0 * 240(lcalpha / DIGIT / "_" / "-" / "*" / "/") "@" lcalpha 0 * 13(lcalpha / DIGIT / "_" / "-" / "*" / "/")
            if (tenantLength == 0 || tenantLength > TraceStateKeyTenantMaxLength)
            {
                return false;
            }

            int vendorLength = key.Length - tenantLength - 1;
            if (vendorLength == 0 || vendorLength > TraceStateKeyVendorMaxLength)
            {
                return false;
            }

            for (int i = tenantLength + 1; i < key.Length; ++i)
            {
                char ch = key[i];
                if (!(IsLowerAlphaDigit(ch)
                    || ch == '_'
                    || ch == '-'
                    || ch == '*'
                    || ch == '/'))
                {
                    return false;
                }
            }

            return true;
        }

        private static bool ValidateValue(string value)
        {
            // https://github.com/w3c/trace-context/blob/master/spec/20-http_request_header_format.md#value
            // value      = 0*255(chr) nblk-chr
            // nblk - chr = % x21 - 2B / % x2D - 3C / % x3E - 7E
            // chr        = % x20 / nblk - chr
            if (value.Length <= 0 || value.Length > TraceStateValueMaxLength)
            {
                return false;
            }

            for (int i = 0; i < value.Length - 1; ++i)
            {
                char c = value[i];
                if (!(c >= 0x20 && c <= 0x7E && c != 0x2C && c != 0x3D))
                {
                    return false;
                }
            }

            char last = value[value.Length - 1];
            return last >= 0x21 && last <= 0x7E && last != 0x2C && last != 0x3D;
        }

        private static bool ValidateValue(ReadOnlySpan<char> value)
        {
            // https://github.com/w3c/trace-context/blob/master/spec/20-http_request_header_format.md#value
            // value      = 0*255(chr) nblk-chr
            // nblk - chr = % x21 - 2B / % x2D - 3C / % x3E - 7E
            // chr        = % x20 / nblk - chr
            if (value.Length <= 0 || value.Length > TraceStateValueMaxLength)
            {
                return false;
            }

            for (int i = 0; i < value.Length - 1; ++i)
            {
                char c = value[i];
                if (!(c >= 0x20 && c <= 0x7E && c != 0x2C && c != 0x3D))
                {
                    return false;
                }
            }

            char last = value[value.Length - 1];
            return last >= 0x21 && last <= 0x7E && last != 0x2C && last != 0x3D;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static bool IsLowerAlphaDigit(char c)
        {
            return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'z');
        }
    }
}

```